### PR TITLE
HDDS-9701. Shareable /tmp dir doesn't work as sticky-bit with Ranger

### PR DESCRIPTION
**Pull request by xBis7**
## What changes were proposed in this pull request?

Ozone native ACLs achieve a sticky-bit behavior for the shareable /tmp dir by checking the key ACLs. Ranger would achieve the same by checking the resource owner and comparing it with the value in the `{OWNER}` tag. For ozone there is no file ownership concept and the user passed to the `RangerOzoneAuthorizer` is always the bucket owner.

If file ownership is implemented, this approach will change but for now we can workaround this issue with a hybrid solution. 

If ACLs aren't native and an external authorizer is defined, we will check the shareable /tmp dir flag. If the shareable /tmp dir is enabled, then we will use a hybrid authorizer. 

The hybrid authorizer checks the resources and if the path is `/tmp/tmp` we will use `OzoneNativeAuthorizer` otherwise we will use the external authorizer.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-9701

## How was this patch tested?

Unit tests were added. The patch was also tested manually with Ranger.

Green CI on my fork: https://github.com/xBis7/ozone/actions/runs/6881630936

### CODE REVIEW

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/security/acl/OzoneAuthorizerFactory.java [hunk 1]
```diff
@@ -78,6 +80,13 @@ private static IAccessAuthorizer create(
...
+    // If authorizer isn't native and shareable tmp dir is enabled,
+    // then return the shared tmp hybrid authorizer.
+    if (conf.getBoolean(OZONE_OM_ENABLE_OFS_SHARED_TMP_DIR,
+        OZONE_OM_ENABLE_OFS_SHARED_TMP_DIR_DEFAULT)) {
+      return new SharedTmpDirAuthorizer(om, conf);
+    }
+
     final IAccessAuthorizer authorizer = newInstance(clazz, conf);
     return authorizer instanceof OzoneNativeAuthorizer
         ? configure((OzoneNativeAuthorizer) authorizer, om, km, pm)
```
Comment (adoroszlai) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/security/acl/OzoneAuthorizerFactory.java:92]: By reording the conditions and passing ready-to-use instances to the hybrid one, we can get rid of the code duplicated in `SharedTmpDirAuthorizer` from this class.

In addition to the suggested change, `configure()` needs to be updated to return `OzoneNativeAuthorizer`.

```suggestion
    final IAccessAuthorizer authorizer = newInstance(clazz, conf);
    
    if (authorizer instanceof OzoneNativeAuthorizer) {
      return configure((OzoneNativeAuthorizer) authorizer, om, km, pm);
    }
    
    // If authorizer isn't native and shareable tmp dir is enabled,
    // then return the shared tmp hybrid authorizer.
    if (conf.getBoolean(OZONE_OM_ENABLE_OFS_SHARED_TMP_DIR,
        OZONE_OM_ENABLE_OFS_SHARED_TMP_DIR_DEFAULT)) {
      return new SharedTmpDirAuthorizer(
        configure(new OzoneNativeAuthorizer(), om, km, pm),
        authorizer);
    }
    
    return authorizer;
```

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/security/acl/SharedTmpDirAuthorizer.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+    Class<? extends IAccessAuthorizer> clazz =
+        conf.getClass(OZONE_ACL_AUTHORIZER_CLASS,
+            OzoneAccessAuthorizer.class,
+            IAccessAuthorizer.class);
+    LOG.info("Initializing class {} from SharedTmpDirAuthorizer", clazz);
+    authorizer = newInstance(clazz, conf);
+  }
+
+  @VisibleForTesting
```
Comment (adoroszlai) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/security/acl/SharedTmpDirAuthorizer.java:60]:
```suggestion
```

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/security/acl/SharedTmpDirAuthorizer.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.apache.hadoop.ozone.security.acl;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.commons.lang.StringUtils;
```
Comment (adoroszlai) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/security/acl/SharedTmpDirAuthorizer.java:20]: Use the one from `commons-lang3`.

```suggestion
import org.apache.commons.lang3.StringUtils;
```

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/security/acl/SharedTmpDirAuthorizer.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
+/**
```
Comment (adoroszlai) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/security/acl/SharedTmpDirAuthorizer.java:1]: nit:

```suggestion
/*
```

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/security/acl/SharedTmpDirAuthorizer.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.apache.hadoop.ozone.security.acl;
+
+import com.google.common.annotations.VisibleForTesting;
```
Comment (adoroszlai) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/security/acl/SharedTmpDirAuthorizer.java:19]:
```suggestion
```

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/security/acl/SharedTmpDirAuthorizer.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+import org.apache.commons.lang.StringUtils;
+import org.apache.hadoop.hdds.conf.OzoneConfiguration;
+import org.apache.hadoop.ozone.om.OzoneManager;
```
Comment (adoroszlai) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/security/acl/SharedTmpDirAuthorizer.java:22]:
```suggestion
```

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/security/acl/SharedTmpDirAuthorizer.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+import org.apache.hadoop.ozone.om.exceptions.OMException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Objects;
+import static org.apache.hadoop.ozone.OzoneConfigKeys.OZONE_ACL_AUTHORIZER_CLASS;
+import static org.apache.hadoop.util.ReflectionUtils.newInstance;
```
Comment (adoroszlai) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/security/acl/SharedTmpDirAuthorizer.java:29]:
```suggestion
```

**Review by adoroszlai — COMMENTED** [review]
Thanks @xBis7 for the patch.

I cannot comment on the Ranger-specific part, but have some suggestions for code improvement.

**Comment by xBis7** [comment]
@adoroszlai Thanks for the improvement suggestions. I've made the changes.

**Comment by xBis7** [comment]
@smengcl @kerneltime Can you take a look at this PR? This is the hybrid authoriser approach that we have discussed before.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/security/acl/SharedTmpDirAuthorizer.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    Objects.requireNonNull(ozObject);
+    Objects.requireNonNull(context);
+    OzoneObjInfo objInfo;
+
+    if (ozObject instanceof OzoneObjInfo) {
+      objInfo = (OzoneObjInfo) ozObject;
+      if (StringUtils.isNotEmpty(objInfo.getVolumeName()) &&
+          StringUtils.isNotEmpty(objInfo.getBucketName()) &&
+          objInfo.getVolumeName().equals("tmp") &&
+          objInfo.getBucketName().equals("tmp")) {
```
Comment (adoroszlai) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/security/acl/SharedTmpDirAuthorizer.java:57]: Let's extract this condition to a method in `OFSPath` and use the constants in that class.

Comment (adoroszlai) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/security/acl/SharedTmpDirAuthorizer.java]: Updated the PR with this.

Comment (xBis7) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/security/acl/SharedTmpDirAuthorizer.java]: @adoroszlai Thanks for updating the PR. Sorry, for not responding, I was on vacation.

**Comment by adoroszlai** [comment]
@smengcl please review

**Comment by adoroszlai** [comment]
@smengcl can you please take a look, or suggest someone who could review this?

**Comment by xBis7** [comment]
This patch doesn't necessarily need someone with Ranger experience, to review it. The purpose of the changes is to bypass an external authorizer when the shareable /tmp is used.

```
if (shareable /tmp enabled) { 
  init HybridAuthorizer
}

HybridAuthorizer
  if (current resource is /tmp/tmp) {
    use OzoneNativeAuthorizer
  } else {
    use external authorizer
  } 
```

If it helps, I can provide more details and an example on why Ranger doesn't work the shareable /tmp.

**Review by adoroszlai — APPROVED** [review]

**Comment by adoroszlai** [comment]
Thanks again @xBis7 for the patch.

**Comment by xBis7** [comment]
@adoroszlai Thanks for reviewing and merging this.

**Pull request merged**